File: templates/create_formula/languages/node/src/formula/output.js
const { existsSync, readFileSync, unlinkSync, writeFileSync } = require("fs")
const { execSync } = require("child_process")
const { homedir } = require('os')

const outputFile = homedir + '/.rit/output.json'

function runFormula(cmd) {
    try {
        execSync(cmd)
    } catch (error) {
        throw error
    }
}

function writeOutput(data) {
    try {
        const dataJSON = JSON.stringify(data)

        writeFileSync(outputFile, dataJSON)
    } catch (error) {
        throw error
    }
}

function readOutput() {
    try {
        if(existsSync(outputFile)) {
            const rawData = readFileSync(outputFile)
            const parseData = JSON.parse(rawData)

            return parseData
        }
    } catch (error) {
        throw error
    }
}

function removeFile() {
    try {
        unlinkSync(outputFile)
    } catch (error) {
        throw error
    }
}

const api = { readOutput, removeFile, runFormula, writeOutput }
module.exports = api

File: templates/create_formula/languages/go/src/pkg/formula/output.go
package formula

import (
	"encoding/json"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type OutputJSON struct {
	Output map[string]interface{} `json:"output"`
}

var outputFile = RitchieDir() + "/output.json"

func RunFormula(command string) error {
	cmdSplit := strings.Split(command, " ")
	cmd := exec.Command(cmdSplit[0], cmdSplit[1:]...)
	if err := cmd.Start(); err != nil {
		return err
	}

	cmd.Wait()

	return nil
}

func WriteOutput(output map[string]interface{}) error {
	data := OutputJSON{output}
	outputJSON, err := json.Marshal(data)
	if err != nil {
		return err
	}

	if err := ioutil.WriteFile(outputFile, outputJSON, os.ModePerm); err != nil {
		return err
	}

	return nil
}

func ReadOutput() (OutputJSON, error) {
	file, _ := ioutil.ReadFile(outputFile)

	var data OutputJSON
	if err := json.Unmarshal(file, &data); err != nil {
		return data, err
	}

	return data, nil
}

func RemoveFile() error {
	if err := os.Remove(outputFile); err != nil {
		return err
	}

	return nil
}

func RitchieDir() string {
	myDir, _ := os.UserHomeDir()

	return filepath.Join(myDir, ".rit")
}
